vars/checkDockerImage.groovy

```diff
@@ -0,0 +1,41 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+/**
+  Checks if the given Docker image exists.
+
+  whenTrue(checkDockerImage(image: 'hello-world:latest')) {
+    ...
+  }
+*/
+def call(Map args = [:]) {
+  def image = args.containsKey('image') ? args.image : error('checkDockerImage: image parameter is required')
+
+  def redirectStdout = isUnix() ? '2>/dev/null' : '2>NUL'
+  if (cmd(returnStatus: true, script: "docker images -q ${image}  ${redirectStdout}") == 0) {
+    log(level: 'DEBUG', text: "${image} exists in the Docker host")
+    return true
+  }
+
+  log(level: 'DEBUG', text: "${image} does not exist: pulling")
+  if (cmd(returnStatus: true, script: "docker pull ${image}") == 0) {
```

Comment:
This is really surprising behavior, tbh. I would not have expected a step which says it will check the status of something to have the side-effect of changing the state of the system.

IMHO this should either be delegated to a step dedicated to pulling an image or it should be put behind a flag which is disabled by default. (Something like `pullIfNotFound: false`. WDYT?

Reply:
In my past experience, check methods were related to check that a state was satisfied, performing the operations to satisfy it if/when needed, that's why I named this step after `check`. In fact I was tempted to create to steps, something like `isDockerImagePulled` and `pullDockerImage` (name them), calling both sequentially. But OTOH I like the name you propose `pullIfNotFound`, as it's more straightforward and clearer to understand its purpose.

Reply:
I meant the name of the attribute (`pullIfNotFound`), although the name of the step should be clearer too

Reply:
@cachedout I added the new parameter, as you suggested. Still looking to a better name for the step, which I'd like to open the discussion.

I like checkXXX because it will try to satisfy the XXX if possible, in contrast to the getter methods (isXXX, getXXX) which simply returns the boolean value after verifying for the current state. In this particular case we have two steps to verify if a Docker image exists:

1. does the image live in the docker host?
2. does the image live in the docker registry?

Thinking out loud, we do not need to effectively pull the image, but simply return YES/NO for 2). I've an idea: using

```shell
docker manifest inspect docker.elastic.co/kibana/kibana:8.0.0-SNAPSHOT > /dev/null 
```

Reply:
With the `docker manifest` approach, I think the name of the step can be changed to `existsDockerImage`. Therefore, we won't need the `pullIfNotFound` attribute.

Reply:
> Still looking to a better name for the step, which I'd like to open the discussion.

`ensureDockerImageExists` ?

Reply:
now that we are not internally pulling, simply querying the docker manifest API, I think `dockerImageExists` is more accurate, as we are not changing the docker host state with a docker-pull command. If you agree, I'm going to rename the step to that name

Reply:
> If you agree, I'm going to rename the step to that name

Yup, makes sense. Thanks!